sentry-android-replay/src/main/java/io/sentry/android/replay/ReplayCache.kt

```diff
@@ -0,0 +1,181 @@
+            MuxerConfig(
+                file = videoFile,
+                recorderConfig = recorderConfig,
+                frameRate = recorderConfig.frameRate.toFloat(),
+                bitrate = 20 * 1000
+            )
+        ).also { it.start() }
+    })
+
+    private val encoderLock = Any()
+    private var encoder: SimpleVideoEncoder? = null
+
+    internal val replayCacheDir: File? by lazy {
+        if (options.cacheDirPath.isNullOrEmpty()) {
+            options.logger.log(
+                WARNING,
+                "SentryOptions.cacheDirPath is not set, session replay is no-op"
+            )
+            null
+        } else {
+            File(options.cacheDirPath!!, "replay_$replayId").also { it.mkdirs() }
+        }
+    }
+
+    // TODO: maybe account for multi-threaded access
+    internal val frames = mutableListOf<ReplayFrame>()
+
+    internal fun addFrame(bitmap: Bitmap, frameTimestamp: Long) {
+        if (replayCacheDir == null) {
+            return
+        }
+
+        val screenshot = File(replayCacheDir, "$frameTimestamp.jpg").also {
+            it.createNewFile()
+        }
+        screenshot.outputStream().use {
+            bitmap.compress(JPEG, 80, it)
+            it.flush()
+        }
+
+        addFrame(screenshot, frameTimestamp)
+    }
+
+    public fun addFrame(screenshot: File, frameTimestamp: Long) {
+        val frame = ReplayFrame(screenshot, frameTimestamp)
+        frames += frame
+    }
+
+    public fun createVideoOf(
+        duration: Long,
+        from: Long,
+        segmentId: Int,
+        videoFile: File = File(replayCacheDir, "$segmentId.mp4")
+    ): GeneratedVideo? {
+        if (frames.isEmpty()) {
+            options.logger.log(
+                DEBUG,
+                "No captured frames, skipping generating a video segment"
+            )
+            return null
+        }
+
+        // TODO: reuse instance of encoder and just change file path to create a different muxer
+        encoder = synchronized(encoderLock) { encoderCreator(videoFile) }
+
+        val step = 1000 / recorderConfig.frameRate.toLong()
+        var frameCount = 0
+        var lastFrame: ReplayFrame = frames.first()
+        for (timestamp in from until (from + (duration)) step step) {
+            val iter = frames.iterator()
+            val frameCountBefore = frameCount
+            while (iter.hasNext()) {
+                val frame = iter.next()
+                if (frame.timestamp in (timestamp..timestamp + step)) {
+                    frameCount++
+                    encode(frame)
+                    lastFrame = frame
+                    break // we only support 1 frame per given interval
+                }
+            }
+
+            // if the frame count hasn't changed we just replicate the last known frame to respect
+            // the video duration.
+            if (frameCountBefore == frameCount) {
+                frameCount++
+                encode(lastFrame)
+            }
```

Comment:
I guess this could be simplified a bit
1. iterate over frames and store matching frame as `lastFrame`
2. exit early (assuming frames are in order)
3. always encode `lastFrame` (either a fresh one, or the one from before) 

```suggestion
            while (iter.hasNext()) {
                val frame = iter.next()
                if (frame.timestamp in (timestamp..timestamp + step)) {
                    lastFrame = frame
                    break // we only support 1 frame per given interval
                }
                // assuming frames are in order, if out of bounds exit early
                if (frame.timestamp < timestamp + step) {
                    break
                }
            }
            frameCount++
            encode(lastFrame)
```

Reply:
I knew this logic looked like something to improve, thanks for this ❤️